dvc/fs/base.py

```diff
@@ -317,25 +317,19 @@ def _download_dir(
         download_files = FsspecCallback.wrap_fn(callback, self._download_file)
         max_workers = jobs or self.jobs
         with ThreadPoolExecutor(max_workers=max_workers) as executor:
-            futures = [
-                executor.submit(download_files, from_info, to_info)
-                for from_info, to_info in zip(from_infos, to_infos)
-            ]
-
             # NOTE: unlike pulling/fetching cache, where we need to
             # download everything we can, not raising an error here might
             # turn very ugly, as the user might think that he has
             # downloaded a complete directory, while having a partial one,
             # which might cause unexpected results in his pipeline.
-            for future in as_completed(futures):
-                # NOTE: executor won't let us raise until all futures that
-                # it has are finished, so we need to cancel them ourselves
-                # before re-raising.
-                exc = future.exception()
-                if exc:
-                    for entry in futures:
-                        entry.cancel()
-                    raise exc
+            try:
+                for _ in executor.imap_unordered(
+                    download_files, from_infos, to_infos, cancel_futures=True
+                ):
+                    pass
+            except Exception:
+                executor.shutdown(wait=False)
+                raise
```

Comment:
We probably don't need this, since we cancel pending futures explicitly, and `ThreadPoolExecutor.__exit__()` will be called with `wait=True` which should be a noop.